### ISSUE TITLE
SSL certificate error at compilation time with dependencies in WSL environment

### ISSUE THREAD

**Issue by AndoneSabin99** [kent-weak-memory/rust#5]
Hello,

I'm interested in studying the compilation process from Rust to CheriLLVM and I come across your compiler which I am trying to use and make it work. Following the setup instructions in the Readme file, I can successfully compile simple programs (without any dependencies) using both cargo and rustc. Also, I'm able to emit MIR and LLVM-IR files for both `aarch64-unknown-freebsd-purecap` and `aarch64-unknown-freebsd` targets.

However, when I attempt to compile programs with dependencies, I encounter the following error when compiling with cargo:

```
error: failed to download from `https://crates.io/api/v1/crates/addr2line/0.13.0/download`

Caused by:
  [60] SSL peer certificate or SSH remote key was not OK (SSL certificate problem: unable to get local issuer certificate)
```

I tried to fix this error by myself, trying several solutions, such as:

1. Putting `http.check-revoke = false` inside the `.cargo/config.toml`
2. Putting `net.git-fetch-with-cli = true` inside the `.cargo/config.toml`
3. Disabling firewalls
4. Not being behind a proxy
5. Updating certificates
6. Rebuilding the compiler from scratch

None of these solutions worked for me. Interestingly, I can access these dependencies with curl without any issues. Also, I tried using another cargo version (the one that is installable via rustup) to check if it can access any crate from crates.io (which it does successfully), so I'm suspecting that this might be actually an issue specific to this cargo version. Either it might be searching for certificates in the wrong directories or overlooking some directories when searching for the certificates.

While I am trying to find a solution (or a workaround) that can solve this problem, I want to ask you if you happened to encounter the same error (or a similar one) and how is it possible to solve it? I would greatly appreciate the assistance.

My environment details:
- WSL Environment: Ubuntu-22.04
- Rust version (the one of this repository):

```
rustc -vV

rustc 1.56.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.56.0-dev
LLVM version: 11.0.0
```

- Cargo version (the one of this repository):

```
cargo -vV

cargo 1.56.0-dev
release: 1.56.0
```

If any additional details are needed, I can gladly provide them.

Many thanks

**Comment by seharris** [comment]
Thanks for the detailed bug report!
I've had a look around to see if I can figure out what's going on here, but I didn't come up with anything you haven't already tried.

We haven't modified Cargo at all, so in theory our version should work exactly the same as the upstream version, other than being a little out of date.
Our version hasn't produced any SSL errors during any of the benchmarking work we've done (that I'm aware of).
Over the course of the project I'd guess that we've compiled at least hundreds of crates from crates.io on both Mac and Linux systems without seeing this kind of problem.
To my knowledge, it's likely that you're the first person to try using our fork under WSL, but I can't immediately see any reason that should break anything.
Given that the version from Rustup works, it's possible that there's been a bug-fix upstream that our version doesn't have yet.
It guess it might be worth seeing if there were any relevant issues closed between 1.56.0 and the current version, it may be possible to back-port a fix if there is one.

For the moment, the best I can say is that we haven't seen this problem before, and I'm quite certain we haven't touched any networking code or bits of Cargo that I think could cause this.
I hope that helps!

**Comment by AndoneSabin99** [comment]
I managed to find a solution to this issue. My initial assumption regarding Cargo overlooking certain directories for certificates was correct. To be sure that this is true, first thing I have done was to run a `strace curl https://crates.io/api/v1/crates/winapi-util/0.1.5/download -v` and check which certificates are being used. 
```
...
openat(AT_FDCWD, "/etc/ssl/certs/ca-certificates.crt", O_RDONLY) = 6
newfstatat(6, "", {st_mode=S_IFREG|0644, st_size=208567, ...}, AT_EMPTY_PATH) = 0
...
```
Here, `curl` uses the ca-certificates.crt in `/etc/ssl/certs/` and has been able to access the url without any issue. Then I did a `strace cargo fetch`, to check in which directory Cargo checks for the certificates. It turns out that Cargo does not look inside the `/etc/ssl/certs/` and instead checks for the certificates in other directories, such as in `/usr/lib/ssl/certs/`.
```
...
openat(AT_FDCWD, "/usr/lib/ssl/certs/ca-certificates.crt", O_RDONLY) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=208567, ...}, AT_EMPTY_PATH) = 0
...
```
Even though the ca-certificates.crt files in both paths have the same certificates, the one that Cargo tries to use does not let it access crates.io for some reason.

The solution to this issue that worked for me has been to specify to Cargo the path to my ca-certificates.crt file, resulting in having something similar like this inside .cargo/config.toml file:
```
[http]
cainfo = "/etc/ssl/certs/ca-certificates.crt"
```
After trying to do `cargo fetch` again, it finally worked and now I am able to compile programs that use dependencies. Furthermore, I can confirm that your compiler works properly under a WSL environment also.
